lib/compat/wordpress-6.5/navigation-block-hooks.php

```diff
@@ -0,0 +1,115 @@
+ *
+ * @param WP_Post $post Post object.
+ */
+function gutenberg_update_ignore_hooked_blocks_meta( $post ) {
+	$anchor_block_type    = 'core/navigation';
+	$first_child_position = 'first_child';
+	$last_child_position  = 'last_child';
+
+	$hooked_blocks                 = get_hooked_blocks();
+	$hooked_block_types_last_child = isset( $hooked_blocks[ $anchor_block_type ][ $last_child_position ] )
+	? $hooked_blocks[ $hooked_blocks[ $anchor_block_type ][ $last_child_position ] ]
+	: array();
+
+	$hooked_block_types_first_child = isset( $hooked_blocks[ $anchor_block_type ][ $last_child_position ] )
+	? $hooked_blocks[ $hooked_blocks[ $anchor_block_type ][ $first_child_position ] ]
+	: array();
+
+	$hooked_block_types_last_child  = apply_filters( 'hooked_block_types', $hooked_block_types_last_child, $last_child_position, $anchor_block_type, null );
+	$hooked_block_types_first_child = apply_filters( 'hooked_block_types', $hooked_block_types_first_child, $first_child_position, $anchor_block_type, null );
+
+	$ignored_hooked_blocks = array_merge( $hooked_block_types_last_child, $hooked_block_types_first_child );
+
+	if ( ! empty( $ignored_hooked_blocks ) && $post->ID ) {
+		update_post_meta( $post->ID, 'ignored_hooked_blocks', json_encode( $ignored_hooked_blocks ) );
+	}
+}
+
+/**
+ * Hooks into the REST API response for the core/navigation block and adds the first and last inner blocks.
+ *
+ * @param WP_REST_Response $response The response object.
+ * @param WP_Post          $post     Post object.
+ * @param WP_REST_Request  $request  Request object.
+ * @return WP_REST_Response $response The response object.
+ */
+function gutenberg_hook_blocks_into_core_navigation_rest_response( $response, $post, $request ) {
+	if ( ! isset( $response->data['content']['raw'] ) || ! isset( $response->data['content']['rendered'] ) ) {
+		return $response;
+	}
+	$parsed_blocks                         = parse_blocks( $response->data['content']['raw'] );
+	$content                               = serialize_blocks( gutenberg_hook_first_last_children( $parsed_blocks, $post->ID ) );
+	$response->data['content']['raw']      = $content;
+	$response->data['content']['rendered'] = apply_filters( 'the_content', $content );
+
+	return $response;
+}
+
+/**
+ * Hooks into the core/navigation block and adds the first and last inner blocks.
+ *
+ * @param WP_Block_List $inner_blocks The inner blocks of the core/navigation block.
+ * @param int|null $post_id Post ID of the navigation block.
+ * @return WP_Block_List $inner_blocks with the first and last inner blocks added.
+ */
+function gutenberg_hook_blocks_into_core_navigation( $inner_blocks, $post_id ) {
+	if ( ! $inner_blocks instanceof WP_Block_List ) {
+		return $inner_blocks;
+	}
+
+	$parsed_blocks = array_map(
+		function ( $inner_block ) {
+			return $inner_block->parsed_block;
+		},
+		iterator_to_array( $inner_blocks )
+	);
```

Comment:
I think you can use [`array_column`](https://www.php.net/manual/en/function.array-column.php) instead:

```suggestion
	$parsed_blocks = array_column(
		iterator_to_array( $inner_blocks ),
		'parsed_block'
	);
```

Come to think of it, do we actually need the `iterator_to_array` conversion? [`class WP_Block_List implements Iterator, ArrayAccess, Countable`](https://github.com/WordPress/wordpress-develop/blob/e707e37f23dcb6723ba3cea22d45f8f35ef95deb/src/wp-includes/class-wp-block-list.php#L15), which I thought should mean that it would work with functions like `array_map` and `array_column` without explicit conversion? 🤔

Reply:
I had that thought too, but using `array_map` directly results in some fatal errors saying it's not an array. I saw `iterator_to_array` was being used [here](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/navigation/index.php#L375-L378) for a similar purpose.

Reply:
Ah, interesting! Oh well, if there's even precedent for this, we'll keep the `iterator_to_array`, I guess 🙂